crates/sui-indexer-alt/src/pipeline/concurrent/collector.rs

```diff
@@ -179,3 +182,166 @@ pub(super) fn collector<H: Handler + 'static>(
         }
     })
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{db, pipeline::Processor};
+
+    use super::*;
+    use prometheus::Registry;
+    use std::time::Duration;
+    use sui_types::full_checkpoint_content::CheckpointData;
+    use tokio::sync::mpsc;
+
+    struct TestHandler;
+    impl Processor for TestHandler {
+        type Value = u64;
+        const NAME: &'static str = "test_handler";
+        const FANOUT: usize = 1;
+
+        fn process(&self, _checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
+            Ok(vec![])
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl Handler for TestHandler {
+        const MIN_EAGER_ROWS: usize = 3;
+        const MAX_CHUNK_ROWS: usize = 4;
+        const MAX_PENDING_ROWS: usize = 10;
+
+        async fn commit(
+            _values: &[Self::Value],
+            _conn: &mut db::Connection<'_>,
+        ) -> anyhow::Result<usize> {
+            tokio::time::sleep(Duration::from_millis(1000)).await;
+            Ok(0)
+        }
+    }
+
+    #[tokio::test]
+    async fn test_collector_batches_data() {
+        let (processor_tx, processor_rx) = mpsc::channel(10);
+        let (collector_tx, mut collector_rx) = mpsc::channel(10);
+        let metrics = Arc::new(IndexerMetrics::new(&Registry::new()));
+        let cancel = CancellationToken::new();
+
+        let _collector = collector::<TestHandler>(
+            PipelineConfig::default(),
+            processor_rx,
+            collector_tx,
+            metrics,
+            cancel.clone(),
+        );
+
+        // Send test data
+        let test_data = vec![
+            Indexed::new(0, 1, 10, 1000, vec![1, 2]),
+            Indexed::new(0, 2, 20, 2000, vec![3, 4]),
+            Indexed::new(0, 3, 30, 3000, vec![5, 6]),
+        ];
+
+        for data in test_data {
+            processor_tx.send(data).await.unwrap();
+        }
+
+        let batch1 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch1.len(), 4);
+
+        let batch2 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch2.len(), 2);
+
+        let batch3 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch3.len(), 0);
```

Comment:
is this expected, why do we want to send empty batches if so? also cc @amnn

Reply:
A batch can have zero rows but still contains watermark updates.